### ISSUE TITLE
[Issue]: Silent wrong results on gfx942 TP8 with GLM-5.2-FP8: degenerate output, fixed by VLLM_ROCM_USE_AITER_CUSTOM_AR=0

### ISSUE THREAD

**Issue by jin-amd** [ROCm/aiter#4855]
### Problem Description

Serving `zai-org/GLM-5.2-FP8` on 8x MI325X (gfx942) with `VLLM_ROCM_USE_AITER=1`,
vLLM selects AITER's `CustomAllreduce` for tensor-parallel all-reduce and enables
AITER's fused allreduce+RMSNorm pass. In that configuration the model intermittently
produces **silently wrong output**: a reply starts coherent and on-topic, then
collapses mid-generation into chains of isolated single digits and never recovers for
the rest of its token budget.
Verbatim example at 30,940 tokens of context, `temperature=0`, HTTP 200, no warning
in the server log:
    Let's dive deep into these four recommendations. If you want to!0,. The1.1.1
    0 2,1,1,1, and5.1.The1a1.... (0,1,1,0,1,0,1,2,0,4,0,0,0,1,1,8,1,1,1,0,1,0,9,...
Setting **`VLLM_ROCM_USE_AITER_CUSTOM_AR=0` and changing nothing else eliminates
it.** Every other AITER kernel (fused MoE, MLA, standalone RMSNorm, fused shared
experts) stays enabled in both arms.

### Operating System

22.04.5 LTS (Jammy Jellyfish)

### CPU

AMD EPYC 9575F 64-Core Processor

### GPU

AMD Instinct MI325X

### ROCm Version

ROCm 7.2.3 

## Environment

| Item | Value |
| --- | --- |
| Hardware | 8x AMD MI325X, `gfx942:sramecc+:xnack-`, single node, fully connected |
| aiter | `amd-aiter 0.1.19` |
| vLLM | `0.26.1rc1.dev942+g5a4c8d992` |
| torch | `2.12.0` |
| Image | `vllm/vllm-openai-rocm:nightly-5a4c8d99242e9e069b604d0e9b969e77f7dd501d` |
| Model | `zai-org/GLM-5.2-FP8`, TP8, no expert parallel |

Two arms, differing by exactly one exported variable.

```bash
export VLLM_WORKER_MULTIPROC_METHOD=spawn
export VLLM_ROCM_USE_AITER=1
export VLLM_ROCM_QUICK_REDUCE_QUANTIZATION=INT4
export VLLM_ROCM_USE_AITER_FUSION_SHARED_EXPERTS=1
export HSA_NO_SCRATCH_RECLAIM=1
# export VLLM_ROCM_USE_AITER_CUSTOM_AR=0     # <-- the entire difference

vllm serve zai-org/GLM-5.2-FP8 \
  --host 0.0.0.0 --port 8000 \
  --block-size 1 \
  --compilation-config '{"pass_config": {"fuse_norm_quant": true, "fuse_act_quant": true}, "custom_ops": ["+rms_norm", "+silu_and_mul"]}' \
  --disable-log-stats --enable-auto-tool-choice --no-enable-expert-parallel \
  --gpu-memory-utilization 0.80 \
  --hf-overrides '{"use_index_cache": true, "index_topk_freq": 8}' \
  --kv-cache-dtype fp8 --max-model-len 263168 \
  --max-num-batched-tokens 16384 --max-num-seqs 128 \
  --no-enable-prefix-caching --reasoning-parser glm45 \
  --tensor-parallel-size 8 --tool-call-parser glm47 --trust-remote-code
```

The server takes 18–41 minutes to become ready on our nodes — 18 and 35 minutes in
the two attached server logs — most of it `torch.compile`. It is not hung. Then:

```bash
BASE_URL=http://localhost:8000 python3 repro_glm52.py --max-context-tokens 70000 > run.log
python3 grade.py --verbose run.log
```

[repro_glm52.py](https://github.com/user-attachments/files/31218553/repro_glm52.py)
[grade.py](https://github.com/user-attachments/files/31218554/grade.py)

### (Optional for Linux users) Output of /opt/rocm/bin/rocminfo --support

<details>
<summary>rocminfo --support output</summary>

```
Paste output here
```

</details>

### Additional Information

The log from today's run is attached here. 
Here are all five, from the stock public nightly. The four turn-57 failures happened in the same scheduler step within a 93-token context window.

conv 0, turn 57, ctx 29,314 — collapses mid-sentence, and note the stray </think> tokens appearing in the garbage:

```
Adding an 8th day to a relaxed itinerary is the ultimate luxury! It means you can
finally cross the river to Osaka's modern north side (Umeda) without rushing...
Here is how to perfectly pace out your extra day (Day 7), pushing your airport
departure to Day !0</think>1.1,0,1,0,1,2, but0.0.0.1 0.1.0.0.0.1 0.1.1,1,1,0,1,0.0.1.1.1.3,0,10,0,1,
```

conv 4, turn 57, ctx 29,387 — breaks after only one clause:

```
Adding a 12th day gives you the ultimate luxury on this route: the ability to!9(1, or5,1,01,1,0.1. The1.1.0.0.1.0.1.0.0
Crime1,0,0,1,1,1,1,1,0,0,0,0,0,1,1,1,1, 0.1.0,0,0,0,1,2,0,0,0,1,84.3.01,...

```
conv 6, turn 57, ctx 29,407 — breaks almost immediately, and contains a Unicode replacement char:

```
If you have 13 days total,!87,0,0,1,1,1,0,1,1,0,0,1,1,1,0,1,1,1,9,0,0, ...1,8,0,1,1,1,1,1,0,0.1.0x0.1.1.1.09.1.0.1.0.1.0
</think>1.1 ... 0,0,0,0,0,:02.11,1,��0,8,
```

conv 1, turn 57, ctx 29,402 — the most severe, degenerating after five words into dense multi-digit runs:
```

If you add a 7!</think>1.1.0,1,0,0,1,0,1,0,1,1,111,1.111:0,1,010.0.1
2.1.0.1.1.0. 1.1.1.0.1.2.2.019.1.1.0112.0,1,1,010 and1.0,011,1,011.5,1,1,021,...

```
conv 2, turn 58, ctx 29,097 — one turn later, and the interesting one: it produces four correct paragraphs of Koh Samui detail before failing mid-sentence at "order!":

```
### **Lunch: The View & The Spice**
*   **The Food:** Around 12:30 PM, order! the1.1.9,1,1,0,1,1,.1.5,0,1, the1, +1,0,1,0,1,1,1,1,1,1,1,1,0, the
11.0.1.1.41
221:0,0,1,0.0.0;0,0,0,0,1,0.0.0.1.51.0.0.0a2.0,...
```

[verify_nn_long70.log](https://github.com/user-attachments/files/31222535/verify_nn_long70.log)

**Comment by jin-amd** [comment]
@TennyWang1223 @xytpai @hubertlu-tw 
Hi all, flagging this one because it produces **silently wrong output** rather than a crash: no NaN, no
assert, HTTP 200, and the request is scored as a normal success.

Short version: on gfx942 TP8 with GLM-5.2-FP8, enabling AITER's custom all-reduce
makes decode output collapse mid-generation into chains of isolated digits.
`VLLM_ROCM_USE_AITER_CUSTOM_AR=0` eliminates it with nothing else changed. We
reproduced it today on the **stock public image**
`vllm/vllm-openai-rocm:nightly-5a4c8d99242e9e069b604d0e9b969e77f7dd501d` with
`amd-aiter 0.1.19` — 5 failures in 1,080 turns, four of them in the *same scheduler
step* across four independent sequences, which is the main reason we suspect a
collective rather than anything per-sequence. A `</think>` token also leaks into
visible output in three of the five, so the logits themselves look corrupted.

We would prefer if any of you could look into this issue. Cause  `VLLM_ROCM_USE_AITER_CUSTOM_AR=0` gives us ~3% perf regression in terms of all matrices. Thanks in advance! If you need any more info regarding this issue, don't hesitate to ping me directly.

**Comment by jin-amd** [comment]
Resolved by [#4346](https://github.com/ROCm/aiter/pull/4346) and https://github.com/ROCm/aiter/pull/4082. Verified on our hardware; we will close it. 

**Verification** (8× MI325X, TP8, `zai-org/GLM-5.2-FP8`, `amd-aiter 0.1.19`):

| aiter | `VLLM_ROCM_USE_AITER_CUSTOM_AR` | Result |
| --- | --- | --- |
| stock | 0 | clean |
| stock | 1 | **corrupts 3/3** |
| +#4346 +#4082 | 1 | **clean** |

Depth run on patched build (`AR=1`, 8 concurrent convs, 200k-token ceiling): **0 corrupt / 2,728 turns** (737 in the 16k–64k band where all 8 original failures lived). Run stopped at 177,776 tokens by external container kill — no traceback, HIP error, or OOM. Same digit-chain detector as the original report; 51,984 body lines scanned, zero hits. `AITER_CUSTOM` confirmed in dispatch with `'fuse_allreduce_rms': True`.

**Root cause:** the fusion half of the `AR=1` gate, not the collective. #4346 adds the missing `end_sync` barrier to fused allreduce+RMSNorm kernels; #4082 (allgather/reduce_scatter) is unused here. Attribution is by elimination — matches the original guess and the race-like signature (silent wrong values, batch-synchronized collapse).

**Repro notes:** (1) Patching the header alone is inert — recompile `module_custom_all_reduce.so`; if using aiter JIT rebuild, pin `pybind11==3.0.1` (pip 3.1.0 breaks cross-module calls). (2) Stock `AR=1` can corrupt on turn 1 (~55 tokens) with 8 concurrent conversations; serial requests do not trigger it.

**Not established:** single host/workload, no gfx950 test, #4346 vs #4082 not isolated experimentally, residual rate not zero-proven.

Re-enabling `VLLM_ROCM_USE_AITER_CUSTOM_AR` (~2.7% output-throughput recovery). Thanks for the fix.
